libbpfgo/libbpf_cb.go

```diff
@@ -10,19 +10,21 @@ import (
 
 //export perfCallback
 func perfCallback(ctx unsafe.Pointer, cpu C.int, data unsafe.Pointer, size C.int) {
-	eventChannels[uintptr(ctx)] <- C.GoBytes(data, size)
+	pb := eventChannels.Get(uint(uintptr(ctx))).(*PerfBuffer)
+	pb.eventsChan <- C.GoBytes(data, size)
 }
 
 //export perfLostCallback
 func perfLostCallback(ctx unsafe.Pointer, cpu C.int, cnt C.ulonglong) {
-	lostChan := lostChannels[uintptr(ctx)]
-	if lostChan != nil {
-		lostChan <- uint64(cnt)
+	pb := eventChannels.Get(uint(uintptr(ctx))).(*PerfBuffer)
+	if pb.lostChan != nil {
+		pb.lostChan <- uint64(cnt)
 	}
 }
 
 //export ringbufferCallback
 func ringbufferCallback(ctx unsafe.Pointer, data unsafe.Pointer, size C.int) C.int {
-	eventChannels[uintptr(ctx)] <- C.GoBytes(data, size)
+	ch := eventChannels.Get(uint(uintptr(ctx))).(chan []byte)
+	ch <- C.GoBytes(data, size)
```

Comment:
Apologies for the confusion, I know i'm likely repeating a question that Yaniv asked but I want to really make sure I understand this. 

Is the whole need for this patch because accessing of the `eventsChannels` map could happen while another ring/perfbuffer is added? If so, wouldn't the newly created ring/perfbuffer not be written as the value of an existing key since the fd will always be unique? Thus avoiding the possibility of this race?

My main point of concern is the above lines in the callback functions. Casting the data as a `*PerfBuffer`/`chan []byte` seems like an expensive operation to be doing every time an event is received. 

We unfortunately don't yet have a robust benchmarking pipeline set up, though we should somewhat soon. I'd feel a lot more comfortable with this change if we could at least use a [go test benchmark](https://golang.org/pkg/testing/#hdr-Benchmarks) to prove that. Does this sound reasonable?

Reply:
Yes, it's because accessing `eventsChannels` map could happen while an insert or delete are happening (when a perf/ring buffer is created or destroyed). The fd's are going to be distinct, of course, but the map (hash table) is the same. It could get re-hashed as a result of an insert. The language specifies it as not thread-safe in this case.

The cast is more of a compile time operation. The type assertion compiles to a compare, a conditional jump and a move. It's just an `if` statement and it'll get very well predicted (since it's a type check). I'm sure the hash table this is replacing has more than a single branch. Not sure if you like reading assembly but I set up a mock to demonstrate: https://godbolt.org/z/rrjxaadv3

I can try to do a go test benchmark if you still have doubts.

Reply:
Ah I believe I follow, thanks for that clarification! I did look at the disassembly before my last comment but didn't find it to be particularly conclusive.

I just read [this article](https://hackernoon.com/some-insights-on-maps-in-golang-rm5v3ywh) which clearly things up greatly for me.

Would this design accomplish the same thing as using a [sync.Map](https://golang.org/pkg/sync/#Map)?

Also I think going back and creating test benchmarks for a ton of libbpfgo functionality would be great for us to do sometime soon but it's fine for now.

Reply:
I forgot about `sync.Map` (I remember seeing it many years ago). Yes, we could just use that and not worry about the capacity limits. I looked at the implementation and since it's more general, it will be a bit slower than what I have but I'm not sure it'll be measurable. Compared to reading data off of perf/ring buffers and doing a channel send, this overhead is probably less than the noise.

Reply:
So do you think we should use the sync.Map instead? I'll rely on your expertise here.

Reply:
re: `sync.Map`: Stability should be the deciding factor when it comes to using it. Can we ensure this map won't grow overtime? IOW, would it be _stable_? There are other factors besides performance like losing compile time type checking by using sync.Map vs a regular mutex+map. I'm not sure if the tradeoff is worth it.

I recommend reading this prior to going down that path: https://github.com/gophercon/2017-talks/blob/master/lightningtalks/BryanCMills-AnOverviewOfSyncMap/An%20Overview%20of%20sync.Map.pdf

Reply:
Thanks for the link to that Simar. My instincts tell me i'd rather tell consumers of libbpfgo to create all perf/ringbuffers up from before accessing (which tracee-ebpf does) than take a performance hit here. I am going to put together some benchmarks to confirm whether or not i'm being paranoid.